DOCSP-44647: add to existing app (#55)
mongodb/docs-mongoid: source/add-existing.txt

.. _mongoid-add-to-existing:

======================================
Add {+odm+} to an Existing Application
======================================

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: ruby framework, odm, migrate

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to add {+odm+} to an existing Sinatra
or Ruby on Rails (Rails) application. To learn how to set up a new
application that uses {+odm+}, see one of the following guides:

- :ref:`mongoid-quick-start-rails`
- :ref:`mongoid-quick-start-sinatra`

Sinatra Application
-------------------

To start using {+odm+} in an existing Sinatra application, you can follow
the same steps described in the
:ref:`Sinatra Quick Start <mongoid-quick-start-sinatra>` guide.

The following steps describe how to add {+odm+} to a Sinatra application:

1. Add the ``mongoid`` dependency to your application's ``Gemfile``.

#. Create a ``config/mongoid.yml`` configuration file and specify your
   connection target.

> #. Create an application file and load your configuration file.

Author: I can link to the example in the QS

#. Create {+odm+} models.

Rails Application
-----------------

You can add {+odm+} to an existing Rails application to run alongside

> other ActiveRecord adapters. To use a combination of adapters, you

Author: I think I'll replace with the spaced version, as I am using it the same way as Mongoid

.. code-block:: ruby
   :caption: Gemfile
   
   gem 'mongoid'

To use {+odm+} as the *only* database adapter, remove or comment out any
RDBMS libraries such as ``sqlite`` or ``pg`` listed in the ``Gemfile``.

Then, install the dependencies by running the following command:

.. code-block:: sh
   
   bundle install

{+odm+} Configuration
~~~~~~~~~~~~~~~~~~~~~

Generate the default {+odm+} configuration by running the following
command:

.. code-block:: sh
   
   bin/rails g mongoid:config

This generator creates the ``config/mongoid.yml`` configuration file
used to configure the connection to the MongoDB deployment and the
``config/initializers/mongoid.rb`` initializer file that you can use
to set other options.

In the ``config/mongoid.yml`` file, specify your connection string and
other connection options.

Modify Frameworks
~~~~~~~~~~~~~~~~~

Open the ``config/application.rb`` file and examine the contents. If it
requires all the components of Rails by including the ``require
'rails/all'`` specification, change this to specify individual
frameworks.

To verify the contents of ``rails/all`` for Rails 7, see the
:github:`rails GitHub repository

> </rails/rails/blob/7-0-stable/railties/lib/rails/all.rb>`.

Author: I actually think we can just delete this note.

The following code is a sample ``config/application.rb`` file that

> demonstrates how to specify individual frameworks instead of using ``rails/all``:

Reviewer: ```suggestion
demonstrates how to load frameworks individually instead of through ``rails/all``:
```

side question: is "frameworks" the right word here? a couple of sources i found use "components" or "libraries" instead.

Author: I see frameworks used in some other references + this was in the original guide that I am adapting

.. code-block:: ruby
   :caption: config/application.rb

   # Remove or comment out rails/all
   #require "rails/all"
 
   # Add the following instead of rails/all:
   require "rails"
 

>    # Comment out unneeded frameworks

Reviewer: solid idea to show how to do this in code. if this is a comprehensive list, and if it isn't too burdensome to keep current, it might be good to direct them here from the opening paragraph ("...as shown in the following code example:")

Author: I'm not sure I totally understand what you mean. This list isn't necessarily what a user would need in their application to use Mongoid, it just demonstrates how to load frameworks individually.

   # require "active_record/railtie" rescue LoadError
   # require "active_storage/engine" rescue LoadError
   require "action_controller/railtie" rescue LoadError
   require "action_view/railtie" rescue LoadError
   require "action_mailer/railtie" rescue LoadError
   require "active_job/railtie" rescue LoadError
   require "action_cable/engine" rescue LoadError
   # require "action_mailbox/engine" rescue LoadError
   # require "action_text/engine" rescue LoadError
   require "rails/test_unit/railtie" rescue LoadError

.. note::

   Because they rely on ActiveRecord, the `ActionText
   <https://guides.rubyonrails.org/action_text_overview.html>`__,

>    `ActiveStorage <https://edgeguides.rubyonrails.org/active_storage_overview.html>`__ and

Reviewer: ```suggestion
   `ActiveStorage <https://edgeguides.rubyonrails.org/active_storage_overview.html>`__, and
```

   `ActionMailbox
   <https://guides.rubyonrails.org/action_mailbox_basics.html>`__
   adapters cannot be used alongside {+odm+}.

Disable ActiveRecord
~~~~~~~~~~~~~~~~~~~~

> Review your application's configuration files, such as
> ``config/application.rb``, then remove or comment out any references to
> ``config.active_record`` and ``config.active_storage``.

Reviewer: S:

```suggestion
In ``config/application.rb`` and your application's other configuration files,
remove or comment out any references to
``config.active_record`` and ``config.active_storage``.
```

Adjust Models
~~~~~~~~~~~~~

> If your application already uses models, you must adjust them to
> migrate from using ActiveRecord to {+odm+}.

Reviewer: Q: do you have to adjust the models in order to migrate? or do you adjust the models such that they can migrate? if it's the former, maybe something like this is clearer:

```suggestion
To migrate from ActiveRecord to {+odm+}, you must adjust your application's models,
if it uses any.
```

> ActiveRecord models derive from ``ApplicationRecord`` and do not have

Reviewer: ```suggestion
ActiveRecord models derive from the ``ApplicationRecord`` class and do not have
```

column definitions. {+odm+} models generally have no superclass but must
include the ``Mongoid::Document`` attribute. {+odm+} models usually
define fields explicitly. You can also use :ref:`dynamic fields

> <mongoid-dynamic-fields>` instead of explicit field definitions.

Reviewer: S: is there a way to tie these four sentences together more? i'm guessing it's explaining how you need to change your models to use Mongoid, but it reads more like a list--in which case, maybe just make it an actual list?

Author: will workshop

For example, a basic ActiveRecord ``Post`` model might resemble the
following:

.. code-block:: ruby
   :caption: app/models/post.rb

   class Post < ApplicationRecord
     has_many :comments, dependent: :destroy
   end

A similar {+odm+} ``Post`` model might resemble the following:

.. code-block:: ruby
   :caption: app/models/post.rb

   class Post
     include Mongoid::Document
 
     field :title, type: String
     field :body, type: String
 
     has_many :comments, dependent: :destroy
   end

> Or, you can define the ``Post`` model by using dynamic fields, as shown

Reviewer: ```suggestion
Instead of using [blank], you can define the ``Post`` model by using dynamic fields, as shown
```

in the following code:

.. code-block:: ruby
   :caption: app/models/post.rb

   class Post
     include Mongoid::Document
     include Mongoid::Attributes::Dynamic
 
     has_many :comments, dependent: :destroy
   end

> {+odm+} does not use ActiveRecord migrations, because MongoDB does not
> require a defined schema before you can store data.

Reviewer: S: maybe my own ignorance, but this feels tacked on to the previous section instead of a logical continuation, so maybe a note admonition would be useful. maybe even move to the next section (or delete altogether), since it makes this same point.

```suggestion
.. note:: ActiveRecord Migrations

   {+odm+} does not use ActiveRecord migrations, because MongoDB does not
   require a defined schema before you can store data.
```

Author: Deleted

-------------------

To start using {+odm+} in an existing Sinatra application, you can follow

> the following steps:

Reviewer: S: change line to "perform the following steps:" or something similar to avoid double "follow"

Modify Frameworks
~~~~~~~~~~~~~~~~~

Open the ``config/application.rb`` file and examine the contents. If the
file uses the ``require "rails/all"`` statement to load all Rails components,
delete this statement. You must add a separate ``require`` statement

> for each Rails component, as shown the following sample

Reviewer: ```suggestion
for each Rails component, as shown in the following sample
```

a predefined schema to store the data.

Migration tools are often specific to datasets.
Even though {+odm+} supports a superset of Active Record associations,
the way that model references are stored in collections is different between

> {+odm+} and Active Record.

Reviewer: S: "{+odm+} and Active Record store model references in collections differently."
